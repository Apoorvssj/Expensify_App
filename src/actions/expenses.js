import { v4 as uuidv4 } from 'uuid';

// ADD_EXPENSE action generator
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuidv4(),
        description,
        note,
        amount,
        createdAt
    }
});

//REMOVE_EXPENSE action generator
//default value is not needed for id, id : id using shorthand
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id 
    
});

//EDIT_EXPENSE action generator
//no need to set defualts,becoz if dont have the id we wont be updating it in the first place
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});