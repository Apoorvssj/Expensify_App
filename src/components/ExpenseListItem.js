

import React from 'react';
import {Link} from 'react-router-dom';


//destructuring props from the expense object,we can also get dispatch from it,
//id : id using shorthand ,passing to removeExpenser
 const ExpenseListItem = ({dispatch, id, description , amount , createdAt}) => (
    <div>
    <Link to={`/edit/${id}`}>
    <h3>Description : {description}</h3>
    </Link>
    <p>Amount: {amount} - CreateD At: {createdAt}</p>
    </div>
);


//export default connect()(ExpenseListItem);



export default ExpenseListItem;