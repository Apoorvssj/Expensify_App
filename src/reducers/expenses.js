//Expenses Reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
      case 'ADD_EXPENSE':
          return [
              //spread operator
              ...state,
              action.expense
          ];
        case 'REMOVE_EXPENSE':
            return state.filter(( {id} ) => id !== action.id); //destructuring state array and taking id from it as id
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        //first adding all properties of expense array then overriding it with new values with spread operator
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense; //no chnage
                }
            });  
        case 'SET_EXPENSES':
            return action.expenses  ;   
      default:
          return state;
  }
};

export default expensesReducer;