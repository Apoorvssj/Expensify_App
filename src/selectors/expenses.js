//startdate and enddate filters changed to moment v13 f11

import moment from 'moment';

// Get visible expenses(filtering expenses)
//destructuring filters
const getVisibleExpenses = (expenses,{ text,sortBy,startDate,endDate }) => {
    //filter and sort both returns an array so we can chain them together to get filtered and sorted array
return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;//it will be always true if there is no startDate,becoz we will never filter things based off of that ,if it is not their(as we have given user ability to clear dates in react-dates in dashboard page)
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch; //if all of these are true the filter will return true and the array(expense) will be kept in expenses array
}).sort((a,b) => {
    if(sortBy === 'date'){
        return a.createdAt < b.createdAt ? 1 : -1; //if this is true 1 will be returned means b will come first,otherwise -1 will be returned means a will come first
    } 
    else if(sortBy === 'amount'){
        return a.amount < b.amount ? 1 : -1;
    }
});
};

export default getVisibleExpenses;

// before moment

/* 

{
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate; // always true for non numbers ,also for undefined .but dont worry we will let undefined filter it
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch; //if all of these are true the filter will return true and the array(expense) will be kept in expenses array
}

*/