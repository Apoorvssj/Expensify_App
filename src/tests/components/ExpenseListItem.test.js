import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem correctly', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});