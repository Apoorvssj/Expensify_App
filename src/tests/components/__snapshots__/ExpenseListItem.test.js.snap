// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Description : 
      Gum
    </h3>
  </Link>
  <p>
    Amount: 
    195
     - CreateD At: 
    0
  </p>
</div>
`;
